r1cs/gadgets/std/src/groups/nonnative/nonnative_group_gadget.rs

```diff
@@ -0,0 +1,1015 @@
+use algebra::{AffineCurve, BitIterator, Field, PrimeField, ProjectiveCurve, SWModelParameters, SquareRootField, curves::short_weierstrass_jacobian::{GroupAffine as SWAffine, GroupProjective as SWProjective}};
+
+use r1cs_core::{ConstraintSystem, SynthesisError};
+
+use crate::{Assignment, ToBitsGadget, ToBytesGadget, alloc::{AllocGadget, ConstantGadget}, boolean::Boolean, fields::{FieldGadget, nonnative::nonnative_field_gadget::NonNativeFieldGadget}, prelude::EqGadget, select::{CondSelectGadget, TwoBitLookupGadget}, uint8::UInt8};
+use std::{borrow::Borrow, marker::PhantomData};
+
+use super::NonNativeGroupGadget;
+
+#[derive(Derivative)]
+#[derivative(Debug, Clone)]
+pub struct GroupAffineNonNativeGadget<
+    P: SWModelParameters<BaseField = SimulationF>,
+    ConstraintF: PrimeField,
+    SimulationF: PrimeField + SquareRootField,
+    F: FieldGadget<SimulationF, ConstraintF>,
+> {
+    pub x:   NonNativeFieldGadget::<SimulationF, ConstraintF>,
+    pub y:   NonNativeFieldGadget::<SimulationF, ConstraintF>,
+    pub infinity:   Boolean,
+    _params: PhantomData<P>,
+    _engine: PhantomData<ConstraintF>,
```

Comment:
PhantomData must be put only if the corresponding template it's not used in any of the members of the struct. Here SimulationF and ConstraintF are used, so you shouldn't need any PhantomData member for them

Reply:
Done